plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java

```diff
@@ -849,6 +868,120 @@ else if (column.isRowPositionColumn()) {
+        int avroSourceChannel = 0;
+
+        Optional<ReaderColumns> columnProjections = projectColumns(columns);
+
+        List<IcebergColumnHandle> readColumns = columnProjections
+                .map(readerColumns -> (List<IcebergColumnHandle>) readerColumns.get().stream().map(IcebergColumnHandle.class::cast).collect(toImmutableList()))
+                .orElse(columns);
+
+        // The column orders in the generated schema might be different from the original order
+        try (DataFileStream<GenericRecord> avroFileReader = new DataFileStream<>(fileIo.newInputFile(path.toString()).newStream(), new GenericDatumReader<>())) {
+            org.apache.avro.Schema avroSchema = avroFileReader.getSchema();
+            List<org.apache.avro.Schema.Field> fileFields = avroSchema.getFields();
+            if (nameMapping.isPresent() && fileFields.stream().noneMatch(IcebergPageSourceProvider::hasId)) {
+                fileFields = fileFields.stream()
+                        .map(field -> setMissingFieldId(field, nameMapping.get(), ImmutableList.of(field.name())))
+                        .collect(toImmutableList());
+            }
+
+            Map<Integer, org.apache.avro.Schema.Field> fileColumnsByIcebergId = mapIdsToAvroFields(fileFields);
+
+            ImmutableList.Builder<String> columnNames = ImmutableList.builder();
+            ImmutableList.Builder<Type> columnTypes = ImmutableList.builder();
+            ImmutableList.Builder<Boolean> rowIndexChannels = ImmutableList.builder();
+
+            for (IcebergColumnHandle column : readColumns) {
+                verify(column.isBaseColumn(), "Column projections must be based from a root column");
+                org.apache.avro.Schema.Field field = fileColumnsByIcebergId.get(column.getId());
+
+                if (column.isPathColumn()) {
+                    constantPopulatingPageSourceBuilder.addConstantColumn(nativeValueToBlock(FILE_PATH.getType(), utf8Slice(path.toString())));
+                }
+                // For delete
+                else if (column.isRowPositionColumn()) {
+                    rowIndexChannels.add(true);
+                    columnNames.add(ROW_POSITION.name());
+                    columnTypes.add(BIGINT);
+                    constantPopulatingPageSourceBuilder.addDelegateColumn(avroSourceChannel);
+                    avroSourceChannel++;
+                }
+                else {
+                    rowIndexChannels.add(false);
+                    columnNames.add(column.getName());
+                    columnTypes.add(column.getType());
+
+                    if (field == null) {
+                        constantPopulatingPageSourceBuilder.addConstantColumn(nativeValueToBlock(column.getType(), null));
+                    }
+                    else {
+                        constantPopulatingPageSourceBuilder.addDelegateColumn(avroSourceChannel);
+                    }
+                    avroSourceChannel++;
```

Comment:
I think these need to be in the else block?
```suggestion
                    if (field == null) {
                        constantPopulatingPageSourceBuilder.addConstantColumn(nativeValueToBlock(column.getType(), null));
                    }
                    else {
                        rowIndexChannels.add(false);
                        columnNames.add(column.getName());
                        columnTypes.add(column.getType());
                        constantPopulatingPageSourceBuilder.addDelegateColumn(avroSourceChannel);
                        avroSourceChannel++;
                    }
```